File: packages/api/src/services/embedding.service.ts
import OpenAI from 'openai';

/**
 * Embedding Service
 *
 * Handles generation of embeddings using OpenAI's API for hybrid retrieval.
 */

// Initialize OpenAI client
const openai = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY,
});

// Embedding model configuration
const EMBEDDING_MODEL = 'text-embedding-3-small';
const EMBEDDING_DIMENSIONS = 1536; // text-embedding-3-small default dimensions

/**
 * Generate embedding for a single query text
 *
 * @param text - Query text to embed
 * @returns Embedding vector as array of numbers
 */
export async function generateQueryEmbedding(text: string): Promise<number[]> {
  if (!text || text.trim() === '') {
    throw new Error('Cannot generate embedding for empty text');
  }

  if (!process.env.OPENAI_API_KEY) {
    throw new Error('OPENAI_API_KEY environment variable not set');
  }

  try {
    const response = await openai.embeddings.create({
      model: EMBEDDING_MODEL,
      input: text.trim(),
      encoding_format: 'float',
    });

    if (!response.data || response.data.length === 0) {
      throw new Error('No embedding returned from OpenAI API');
    }

    const embedding = response.data[0].embedding;

    if (!embedding || embedding.length === 0) {
      throw new Error('Invalid embedding returned from OpenAI API');
    }

    return embedding;
  } catch (error) {
    console.error('Failed to generate embedding:', error);
    throw new Error(`Embedding generation failed: ${error instanceof Error ? error.message : 'Unknown error'}`);
  }
}

/**
 * Generate embeddings for multiple texts (batch processing)
 *
 * @param texts - Array of texts to embed
 * @returns Array of embedding vectors
 */
export async function generateEmbeddingsBatch(texts: string[]): Promise<number[][]> {
  if (!texts || texts.length === 0) {
    return [];
  }

  if (!process.env.OPENAI_API_KEY) {
    throw new Error('OPENAI_API_KEY environment variable not set');
  }

  // Filter out empty texts
  const validTexts = texts.filter(t => t && t.trim() !== '');

  if (validTexts.length === 0) {
    return [];
  }

  try {
    const response = await openai.embeddings.create({
      model: EMBEDDING_MODEL,
      input: validTexts.map(t => t.trim()),
      encoding_format: 'float',
    });

    if (!response.data || response.data.length === 0) {
      throw new Error('No embeddings returned from OpenAI API');
    }

    // OpenAI returns embeddings in the same order as input
    return response.data.map(item => item.embedding);
  } catch (error) {
    console.error('Failed to generate embeddings batch:', error);
    throw new Error(`Batch embedding generation failed: ${error instanceof Error ? error.message : 'Unknown error'}`);
  }
}

/**
 * Get embedding model info
 */
export function getEmbeddingModelInfo() {
  return {
    model: EMBEDDING_MODEL,
    dimensions: EMBEDDING_DIMENSIONS,
  };
}

File: packages/workers/src/services/embedding_service.py
"""Embedding service supporting multiple providers (OpenAI, Gemini).

The provider is selected via the EMBEDDING_PROVIDER env var (default: "openai").
Model and dimension can be overridden via EMBEDDING_MODEL / EMBEDDING_DIMENSION.
"""
import logging
from typing import List

from ..config import settings

logger = logging.getLogger(__name__)

# Provider-specific defaults
_DEFAULTS = {
    "openai": {"model": "text-embedding-3-small", "dimension": 1536},
    "gemini": {"model": "models/text-embedding-004", "dimension": 768},
}


def _resolve(provider: str):
    """Return (model, dimension) honoring config overrides."""
    defaults = _DEFAULTS.get(provider, _DEFAULTS["openai"])
    model = settings.embedding_model or defaults["model"]
    dimension = settings.embedding_dimension or defaults["dimension"]
    return model, dimension


class EmbeddingService:
    """Generates semantic embeddings via the configured provider."""

    def __init__(self):
        self.provider = (settings.embedding_provider or "openai").lower()
        self.model, self.dimension = _resolve(self.provider)

        if self.provider == "gemini":
            key = settings.gemini_api_key
            if key:
                logger.info("GEMINI_API_KEY found (%s…%s)", key[:4], key[-4:])
            else:
                logger.error("GEMINI_API_KEY is NOT set — embedding calls will fail!")
            import google.generativeai as genai
            genai.configure(api_key=key)
            self._genai = genai
        elif self.provider == "openai":
            key = settings.openai_api_key
            if key:
                logger.info("OPENAI_API_KEY found (%s…%s)", key[:4], key[-4:])
            else:
                logger.error("OPENAI_API_KEY is NOT set — embedding calls will fail!")
            from openai import OpenAI
            self._openai = OpenAI(api_key=key)
        else:
            raise ValueError(f"Unsupported EMBEDDING_PROVIDER: {self.provider!r}")

        # Use print() so this shows even before logging is configured
        print(f"[config] Provider initialized: provider={self.provider} model={self.model} dimension={self.dimension}")
        logger.info(
            "EmbeddingService initialised: provider=%s  model=%s  dimension=%d",
            self.provider, self.model, self.dimension,
        )

    # ------------------------------------------------------------------

    def embed_query(self, text: str) -> List[float]:
        """Generate an embedding for a single search query."""
        logger.info("Generating embedding for query (length=%d chars)", len(text))
        if self.provider == "gemini":
            embedding = self._embed_gemini([text])[0]
        else:
            embedding = self._embed_openai([text])[0]
        logger.info("Embedding generated: dimension=%d", len(embedding))
        return embedding

    def embed_texts(self, texts: List[str]) -> List[List[float]]:
        """Batch embed multiple texts (for ingestion)."""
        logger.info("Generating embeddings for %d texts", len(texts))
        if self.provider == "gemini":
            embeddings = self._embed_gemini(texts)
        else:
            embeddings = self._embed_openai(texts)
        logger.info("Generated %d embeddings", len(embeddings))
        return embeddings

    # ------------------------------------------------------------------
    # Provider implementations
    # ------------------------------------------------------------------

    def _embed_openai(self, texts: List[str]) -> List[List[float]]:
        response = self._openai.embeddings.create(model=self.model, input=texts)
        return [item.embedding for item in response.data]

    def _embed_gemini(self, texts: List[str]) -> List[List[float]]:
        result = self._genai.embed_content(
            model=self.model,
            content=texts,
            output_dimensionality=self.dimension,
        )
        # embed_content returns {"embedding": [...]} for single text
        # and {"embedding": [[...], ...]} for a list of texts.
        emb = result["embedding"]
        if texts and isinstance(emb[0], float):
            return [emb]
        return emb
